backend/drm/drm.c

```diff
@@ -1613,3 +1624,141 @@ void destroy_drm_connector(struct wlr_drm_connector *conn) {
 	wl_list_remove(&conn->link);
 	free(conn);
 }
+
+int wlr_drm_backend_get_read_only_fd(struct wlr_backend *backend) {
+	assert(backend);
+
+	struct wlr_drm_backend *drm = get_drm_backend_from_backend(backend);
+	char *path = drmGetDeviceNameFromFd2(drm->fd);
+	if (!path) {
+		wlr_log(WLR_ERROR, "Failed to get device name from DRM fd");
+		return -1;
+	}
+
+	int fd = open(path, O_RDWR);
```

Comment:
Why is `wlr_drm_backend_get_read_only_fd` opening the fd `O_RDWR`? Am I missing something?

Also, why is this open not done through `wlr_session`? Is the user expected to be in the `video` group for this?

Reply:
You can't perform any IOCTL unless you've opened the FD with `O_RDWR`.

> Also, why is this open not done through wlr_session?

logind can't re-open devices. See https://github.com/systemd/systemd/issues/19512.

> Is the user expected to be in the video group for this?

Sadly, on non-systemd setups, yes. The long-term plan is to introduce read-only DRM leases.

Reply:
Maybe we could try opening the device with the session, and fallback to plain `open` if that fails? That way, things would work as expected on non-systemd setups, without having to change the card node permissions.

Reply:
> You can't perform any IOCTL unless you've opened the FD with O_RDWR.

That's fair, but I see nothing making this read only as the name implies. Seems like the function returns a normal writable DRM FD despite its name.

> Sadly, on non-systemd setups, yes. The long-term plan is to introduce read-only DRM leases.

On systemd, will uaccess be used to grant the permissions instead of TakeDevice? Otherwise it will also be a problem on systemd.

Maybe we need a uaccessd for non-systemd, but I never liked that thing...

> Maybe we could try opening the device with the session, and fallback to plain open if that fails?

seatd currently dups if you open the same file again, as I didn't know what to do about drmSetMaster if you opened the same thing twice. This is easy to change if we know what behavior we want.

Reply:
> That's fair, but I see nothing making this read only as the name implies. Seems like the function returns a normal writable DRM FD despite its name.

The intention is to provide a DRM FD which can only be used to query information, and can't be used to change the KMS configuration. Maybe `non_master` would be a better name?

> On systemd, will uaccess be used to grant the permissions instead of TakeDevice? Otherwise it will also be a problem on systemd.

Hm, what do you mean? Won't uaccess set the ACLs on the file?

> seatd currently dups if you open the same file again, as I didn't know what to do about drmSetMaster if you opened the same thing twice. This is easy to change if we know what behavior we want.

Ah yeah, we absolutely can't use a dup'ed FD, because everything KMS does is per file description, not per file descriptor. So `drmDropMaster` will also drop master for our internal DRM FD.